Change NotifyOnChangeOnly to accept a list

Beastlybear2017 opened the pull request:
I apologise if this is really dumb, I have never attempted to contribute towards a npm package before and have no clue how to / if it is possible to test it from a local file.

Here is my vision of how this would work, although have no clue if it works in practice :)

Beastlybear2017 committed:
Update package.json

Beastlybear2017 committed:
Update Parser.ts

Beastlybear2017 committed:
Update package.json

Beastlybear2017 committed:
Update Parser.ts

Beastlybear2017 committed:
Update package.json

Beastlybear2017 committed:
Update package.json

Beastlybear2017 committed:
Update package.json

ghaiklor commented:
@Beastlybear2017 you can try to write a test, where to check if your solutions works.

Here, for instance, test case that checks if the `notifyChange` works:

https://github.com/ghaiklor/icecast-parser/blob/e6ba7a3cb7161eb0c9ce08039533fa5f8afa1dda/test/Parser.spec.ts#L57-L65

You can add more cases like that where you check scenarios like “what if you provided no keys at all, that way it should notify every time” or “what if I provided the keys that doesn't exist in metadata, it should notify” and so on.

Beastlybear2017 committed:
Update Parser.spec.ts

Beastlybear2017 committed:
Update Parser.spec.ts

Beastlybear2017 committed:
Update Parser.ts

Beastlybear2017 committed:
Update Parser.ts

Beastlybear2017 committed:
Update Parser.spec.ts

Beastlybear2017 commented:
I have added the tests (updated yours to use the new system), fixed any issues and commited my changes. All the tests pass / fail when they should
https://github.com/Beastlybear2017/icecast-parser/blob/ccab52975b0fdf5586ff5734e9a171cdcffe779a/test/Parser.spec.ts#L68-L95

![image](https://github.com/ghaiklor/icecast-parser/assets/96798368/48f2cc42-3e90-4841-90d9-646514e898ec)

Beastlybear2017 committed:
Update Parser.ts

Fixed previous issue - Now works as intened

Beastlybear2017 commented:
Was able to test within my project and fixed an issue that I discorvered. Now works as intened

Beastlybear2017 committed:
Update README.md

Updates info for notifyOnChangeOnly

ghaiklor commented:
Great work!

However, let's keep it backward compatible with older API. Let's leave the possibility to specify the boolean as before.

ghaiklor commented on README.md:
```diff
@@ -47,7 +47,7 @@ You can provide additional parameters to constructor:
 - `keepListen` - by default `false`. If you set to `true`, then response from radio station will not be destroyed and you can pipe it to another streams. E.g. piping it to the `speaker` module.
 - `autoUpdate` - by default `true`. If you set to `false`, then parser will not be listening for recent updates and immediately close the stream. So that, you will get a metadata only once.
-- `notifyOnChangeOnly` - by default `false`. If you set both `autoUpdate` and `notifyOnChangeOnly` to `true`, it will keep listening the stream and notifying you about metadata, but it will not notify if metadata did not change from the previous time.
+- `notifyOnChangeOnly` - by default `[]`. If you set both `autoUpdate` to `true` and `notifyOnChangeOnly` to include the a list of metadata keys to listen for (`['StreamTitle', 'StreamUrl']`), it will keep listening the stream and notifying you about metadata, but it will not notify if the selected metadata did not change from the previous time.
```
By default, let it be `false` as before for backward compatibility.

ghaiklor commented on src/Parser.ts:
```diff
@@ -9,7 +9,7 @@ export interface ParserOptions {
   keepListen: boolean
   metadataInterval: number
-  notifyOnChangeOnly: boolean
+  notifyOnChangeOnly: string[]
```
```suggestion
  notifyOnChangeOnly: boolean | string[]
```

ghaiklor commented on src/Parser.ts:
```diff
@@ -35,7 +35,7 @@ export class Parser extends EventEmitter {
     keepListen: false,
     metadataInterval: 5,
-    notifyOnChangeOnly: false,
+    notifyOnChangeOnly: [],
```
```suggestion
    notifyOnChangeOnly: false,
```

ghaiklor commented on src/Parser.ts:
```diff
@@ -115,12 +115,22 @@ export class Parser extends EventEmitter {
-    for (const [key, value] of metadata.entries()) {
-      if (this.previousMetadata.get(key) !== value) {
-        return true;
+    if (this.options.notifyOnChangeOnly.length > 0) {
```
Now, here we can check the type of the option. If there is an array, we treat it by your logic but if there is a boolean, nothing changes and works as before.

ghaiklor replied:
```ts
if (Array.isArray(this.options.notifyOnChangeOnly)) {
  // your new logic
} else if (this.options.notifyOnChangeOnly === true) {
  // old logic
}
```

ghaiklor commented on test/Parser.spec.ts:
```diff
@@ -57,11 +57,40 @@ describe('parser', () => {
     expect.hasAssertions();
 
-    const radio = new Parser({ autoUpdate: false, notifyOnChangeOnly: true, url: 'https://live.hunter.fm/80s_high' });
+    const radio = new Parser({ autoUpdate: false, notifyOnChangeOnly: [], url: 'https://live.hunter.fm/80s_high' });
```
This change is no longer needed then and this test will check that everything works as before
